drivers/hwspinlock/sqn_hwspinlock.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2023 Sequans Communications
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT sqn_hwspinlock
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/sys_io.h>
+#include <zephyr/drivers/hwspinlock.h>
+
+#include <zephyr/sys/printk.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sqn_hwspinlock);
+
+struct sqn_hwspinlock_data {
+	DEVICE_MMIO_RAM;
+	uint32_t num_locks;
+};
+
+struct sqn_hwspinlock_config {
+	DEVICE_MMIO_ROM;
+};
+
+static inline mem_addr_t get_lock_addr(const struct device *dev, uint32_t id)
+{
+	return (mem_addr_t)(DEVICE_MMIO_GET(dev) + id * sizeof(uint32_t));
+}
+
+/*
+ * To define CPU id, we use the affinity2 and affinity1
+ * fields of the MPIDR register.
+ */
+static uint8_t mpidr_to_cpuid(uint64_t mpidr_val)
+{
+	uint8_t cpuid = ((mpidr_val >> 8) & 0x0F) | ((mpidr_val >> 12) & 0xF0);
+
+	return ++cpuid;
+}
+
+static int sqn_hwspinlock_trylock(const struct device *dev, uint32_t id)
+{
+	struct sqn_hwspinlock_data *data = dev->data;
+	uint8_t cpuid;
+
+	if (id > data->num_locks)
+		return -EINVAL;
```

Comment:
should we add a new `_get_numlock` hook in the API to retrieve the ID max?

This to not have to return `int` from `lock / unlock` and still be sure to not pass a wrong ID.

Reply:
Good idea, but I don't know how to add it... I don't want to add an API just for that.
Do you have an idea?

Reply:
Why not adding `hwspinlock_api_get_max_id` to the API you introduced in the first commit?

Reply:
Done.